homeassistant/components/sensor/tof.py

```diff
@@ -0,0 +1,135 @@
+https://github.com/josemotta/vl53l1x-python
+https://github.com/NagyAttila/vl53l1x-python
+https://github.com/pimoroni/vl53l1x-python
+
+- DEFAULT_RANGE is always LONG = 2, for 4 meters range.
+- A GPIO connected to VL53L1X XSHUT input resets the device.
+- XSHUT starts pulsing LOW and after that it is kept HIGH all time.
+
+"""
+import asyncio
+import logging
+from functools import partial
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components import rpi_gpio
+from homeassistant.const import CONF_NAME
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['smbus2==0.2.2', 'VL53L1X2==0.1.5']
+
+DEPENDENCIES = ['rpi_gpio']
+
+_LOGGER = logging.getLogger(__name__)
+
+LENGTH_MILLIMETERS = 'mm'
+
+CONF_I2C_ADDRESS = 'i2c_address'
+CONF_I2C_BUS = 'i2c_bus'
+CONF_XSHUT = 'xshut'
+
+DEFAULT_NAME = 'VL53L1X'
+DEFAULT_I2C_ADDRESS = 0x29
+DEFAULT_I2C_BUS = 1
+DEFAULT_XSHUT = 16
+DEFAULT_RANGE = 2
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME,
+                 default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_I2C_ADDRESS,
+                 default=DEFAULT_I2C_ADDRESS): vol.Coerce(int),
+    vol.Optional(CONF_I2C_BUS,
+                 default=DEFAULT_I2C_BUS): vol.Coerce(int),
+    vol.Optional(CONF_XSHUT,
+                 default=DEFAULT_XSHUT): cv.positive_int,
+})
+
+
+def init_tof(xshut, level):
+    """XSHUT port LOW resets the device."""
+    rpi_gpio.setup_output(xshut)
+    rpi_gpio.write_output(xshut, level)
+
+
+async def async_setup_platform(hass,
+                               config,
+                               async_add_entities,
+                               discovery_info=None):
+    """Reset and initialize the VL53L1X ToF Sensor from STMicroelectronics."""
+    from VL53L1X2 import VL53L1X  # pylint: disable=import-error
+
+    name = config.get(CONF_NAME)
+    bus_number = config.get(CONF_I2C_BUS)
+    i2c_address = config.get(CONF_I2C_ADDRESS)
+    unit = LENGTH_MILLIMETERS
+    xshut = config.get(CONF_XSHUT)
+
+    await hass.async_add_executor_job(
+        init_tof, xshut, 0
+    )
+    await asyncio.sleep(0.01)
+    await hass.async_add_executor_job(
+        init_tof, xshut, 1
+    )
+    await asyncio.sleep(0.01)
+
+    sensor = await hass.async_add_executor_job(
+        partial(VL53L1X, bus_number)
+    )
+
+    dev = [VL53L1XSensor(sensor, name, unit, i2c_address)]
+
+    async_add_entities(dev, True)
+
+
+class VL53L1XSensor(Entity):
+    """Implementation of VL53L1X sensor."""
+
+    def __init__(self, vl53l1x_sensor, name, unit, i2c_address):
+        """Initialize the sensor."""
+        self._name = name
+        self._unit_of_measurement = unit
+        self.vl53l1x_sensor = vl53l1x_sensor
+        self.i2c_address = i2c_address
+        self._state = None
+        self.vl53l1x_sensor.open()
```

Comment:
Does it do I/O?

Reply:
Maybe, I checked the C driver and looks like it does a software reset on the device. But __init__ is not async! Should I use async __init__?

Reply:
No. Then we need to move it out of init.

Reply:
We can move it back inside async_update, but in this case, the device will be initialized again at every update. Not a problem for slow applications. But the device itself is very fast, would support measurements up to 50 Hz. Maybe for version 2.0!

Reply:
Done!

```
class VL53L1XSensor(Entity):
    """Implementation of VL53L1X sensor."""

    def __init__(self, vl53l1x_sensor, name, unit, i2c_address):
        """Initialize the sensor."""
        self._name = name
        self._unit_of_measurement = unit
        self.vl53l1x_sensor = vl53l1x_sensor
        self.i2c_address = i2c_address
        self._state = None
...
    def measure(self):
        """Get the latest measurement from VL53L1X."""
        self.vl53l1x_sensor.open()
        self.vl53l1x_sensor.add_sensor(self.i2c_address, self.i2c_address)
        self.vl53l1x_sensor.start_ranging(self.i2c_address, DEFAULT_RANGE)
        self.vl53l1x_sensor.update(self.i2c_address)
        self.vl53l1x_sensor.stop_ranging(self.i2c_address)

    async def async_update(self):
        """Get the latest measurement and update state."""

        await self.hass.async_add_executor_job(
            self.measure
        )
```